platform/linux-generic/arch/aarch64/odp_random.h

```diff
@@ -0,0 +1,172 @@
+/* Copyright (c) 2021, ARM Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+#ifndef ODP_AARCH64_RANDOM_H_
+#define ODP_AARCH64_RANDOM_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <odp/api/spec/random.h>
+#include <odp/autoheader_internal.h>
+
+#include <stdint.h>
+
+odp_random_kind_t _odp_random_max_kind_generic(void);
+int32_t _odp_random_true_data_generic(uint8_t *buf, uint32_t len);
+int32_t _odp_random_crypto_data_generic(uint8_t *buf, uint32_t len);
+
+#ifdef __ARM_FEATURE_RNG
+
+static inline int _odp_random_max_kind(void)
+{
+	return ODP_RANDOM_TRUE;
+}
+
+static inline int _odp_rndr(unsigned long *v)
+{
+	int pass;
+
+	/* Return a 64-bit random number which is reseeded from the True Random
+	 * Number source. If the hardware returns a genuine random number,
+	 * PSTATE.NZCV is set to 0b0000. The NZCV condition flag is checked via
+	 * the CSET instruction. If the hardware cannot return a genuine random
+	 * number in a reasonable period of time, PSTATE.NZCV is set to 0b0100
+	 * and the data value returned is 0. */
+	__asm__ volatile("mrs   %0, s3_3_c2_c4_0\n\t"
+			 "cset  %w[pass], ne"
+			 : "=r" (*v), [pass] "=r" (pass)
+			 :
+			 : "cc");
+
+	return pass;
+}
+
+static inline int _odp_rndrrs(unsigned long *v)
+{
+	int pass;
+
+	/* Return a 64-bit random number which is reseeded from the True Random
+	 * Number source immediately before the read of the random number.
+	 * If the hardware returns a genuine random number, PSTATE.NZCV is
+	 * set to 0b0000. The NZCV condition flag is checked via the CSET
+	 * instruction. If the hardware cannot return a genuine random number
+	 * in a reasonable period of time, PSTATE.NZCV is set to 0b0100 and the
+	 * data value returned is 0. */
+	__asm__ volatile("mrs   %0, s3_3_c2_c4_1\n\t"
+			 "cset  %w[pass], ne"
+			 : "=r" (*v), [pass] "=r" (pass)
+			 :
+			 : "cc");
+
+	return pass;
+}
+
+static inline int32_t _odp_random_crypto_data(uint8_t *buf, uint32_t len)
+{
+	for (uint32_t i = 0; i < len / 8; i++) {
+		uint64_t temp;
+
+		while (!_odp_rndr(&temp))
+			;
+		*((uint64_t *)buf) = temp;
```

Comment:
buf may be unaligned. One possibility would be to test for __ARM_FEATURE_UNALIGNED in addition to __ARM_FEATURE_RNG, i.e. `#if defined(__ARM_FEATURE_RNG) && defined(__ARM_FEATURE_UNALIGNED)` at the top of the file. Maybe @psavol can comment if that's sufficient here.

Reply:
I guess __ARM_FEATURE_UNALIGNED should be supported on ARMv8 always. So, it would be sufficient to just stop the build if RNG is supported, but UNALIGNED not.

```
#ifdef __ARM_FEATURE_RNG

#if __ARM_FEATURE_UNALIGNED != 1
#error Unaligned access not supported
#endif
```

Reply:
Sounds good, I will make the change. Thanks!